### ISSUE TITLE
prisma .Disconnect() does not work properly

### ISSUE THREAD

**Issue by steebchen** [steebchen/prisma-client-go#545]
When an OS signal is sent, the query engine is immediately killed, so ongoing queries can't finish. The query engine should only be killed once .Disconnect() is called.

**Comment by steebchen** [comment]
@matthewmueller Do you mean this https://github.com/prisma/prisma/issues/2917?

**Comment by steebchen** [comment]
Some observations:

- Right now all signals are forwarded to the query engine, as it is the default behaviour when spawning a child process. This means when a docker container is stopped or a user presses CTRL+C in the console, SIGINT is sent to the parent process (in this case Go), but also to the QE at the same time, as the signal is forwarded.
- When a user intercepts the signal in the parent process, they may clean up things, but if database queries are still running, they are killed, as the query engine does not seem to check if queries are still running.
- This is why the current implementation is broken, and when a user wants to gracefully exit their app including database queries, they can't.
- I suggest to stop forwarding signals via `Setpgid`, so that signals sent to the parent process will NOT be forwarded to the query engine.
- This means the user would need to listen for signals, then call Disconnect() manually after they make sure no queries are running anymore (e.g. by gracefully closing their http server, so that when all http requests are closed, it's almost certain that no queries are running anymore), then call prisma.Disconnect() to notify the query engine (which actually sends a SIGINT to the child process internally), and only then exit the process. It seems this is the only way to have a fully working graceful exit of a process with prisma.
- A disadvantage of this approach is that when a user forgets to call prisma.Disconnect(), the query engine will continue to run in the background until manually killed or a machine restart.
- Regardless, it seems that if a process is killed forcefully (SIGKILL), the query engine will not be killed either way.

Alternatives:
- Add a config option whether to forward signals or not, something like "AutoDisconnect: bool"
- The query engine could implement its own mechanism of existing once the parent process exits:
  - One option is PR_SET_PDEATHSIG, but I don't understand how that works or if that would solve the issue
  - Monitor actively running queries (or rather connected http clients when using the internal http engine) and if none are running/connected, exit. For non-http protocols like NAPI only watching queries would work.
  - One caveat: When the user wants to finish a process (like processing a MQ item which may do multiple queries), this my not work with NAPI, because the parent process runs longer and the child would need to wait for the parent to finish first, as watching queries would not be sufficient.

**Comment by steebchen** [comment]
Proposed changes: #555

**Comment by steebchen** [comment]
TLDR: This works as intended. The behaviour just seems to differ in terminals, because they send signals to both parent and child processes, whereas if you want to gracefully exit a process, a signal is just sent to the parent, which then can gracefully shutdown whatever it does and then call prisma.Disconnect() to gracefully shut down the query engine.

**Comment by steebchen** [comment]
From the notes above, the best solution actually seems to be using setpgid: true. The user simply has to call .Disconnect(). Optionally I should potentially add an option to toggle of this behaviour, it might create problems with auto restart dev apps.

**Comment by steebchen** [comment]
https://github.com/steebchen/prisma-client-go/blob/c207dc03a144cd310cdffd9cd50b8598a4a211b4/engine/lifecycle_unix.go#L9
